### DISCUSSION TITLE
Long-living effect memory issue when using concurrency

### DISCUSSION THREAD

**Discussion by ikarishinjigao** [pointfreeco/swift-composable-architecture#1337] [Q&A] [answered]
Hi, I'm trying to use the TCA (which is awesome BTW) to implement some long-living effect to change state with high-frequency actions, but here is some memory issue i guess.

I'm using the AVFoundation to get CVPixelBuffer and send an action to update state, in order to produce the issue here, for example, if I change the [long-living example here](https://github.com/pointfreeco/swift-composable-architecture/blob/108e3a536fcebb16c4f247ef92c2d7326baf9fe3/Examples/CaseStudies/SwiftUICaseStudies/02-Effects-LongLiving.swift#L42-L44) to the following,

```swift
// Simulate high-frequency action
for await _ in await DispatchQueue.main.timer(interval: 0.01) {
  await send(.userDidTakeScreenshotNotification)
}
```

the memory will keeps increasing.

![Screen Shot 2022-09-06 at 2 54 07](https://user-images.githubusercontent.com/9335999/188498390-49b35a61-a992-400a-a235-32085aebe3d7.png)

I can confirm that [the `tasks` defined here](https://github.com/pointfreeco/swift-composable-architecture/blob/cfdb845e5707191f87d0e8c72180d4d46c4f4158/Sources/ComposableArchitecture/Store.swift#L335) keeps increasing like this, and not being cleaned up. (I don't know much about what exactly going on here...)

![Screen Shot 2022-09-06 at 3 06 52](https://user-images.githubusercontent.com/9335999/188498742-978a0006-bb8e-4099-adb8-6fcc10ca71b5.png)

According to [the doc](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance#High-frequency-actions), I understand that the TCA is not designed or not recommended for high-frequency actions, but what I facing seems a bug to me, and couldn't find any doc or issue about this one.

Am I doing something wrong? Can someone help me with this?
Sorry for my english, I'm not a native speaker... :(

Tested on Xcode 14.0 beta 6, iPhone 13 Pro Max (Simulator), main branch.

**Comment by stephencelis** [comment, marked as answer]
Good catch! There was definitely a leak here. Tasks are so small that it takes some time and speed for the leak to build up, but it's definitely there. I've pushed #1338 to address.

**Reply by ikarishinjigao** [reply]
Thanks! Issue fixed!
